src/main.rs

```diff
@@ -42,15 +49,139 @@ fn read_can(mut publisher: MqttClient, can_interface: &str) {
         );
         let decoded_data = message.decode();
         for data in decoded_data.iter() {
-            publisher.publish(data)
+            let mut payload = serverdata::ServerData::new();
+            payload.unit = data.unit.to_string();
+            payload.value = data.value.iter().map(|x| x.to_string()).collect();
+
+            client
+                .publish(
+                    data.topic.to_string(),
+                    protobuf::Message::write_to_bytes(&payload).unwrap_or_else(|e| {
+                        format!("failed to serialize {}", e).as_bytes().to_vec()
+                    }),
+                )
+                .expect("Could not publish!");
+            // TODO: investigate disabling this
+            thread::sleep(Duration::from_micros(100));
         }
+    });
+    Ok(join_handle)
+}
+
+/**
+ * Reads the mqtt incoming messages and sends can messages based off of that
+ */
+fn read_siren(
+    pub_path: &str,
+    send_map: Arc<RwLock<HashMap<u32, EncodeData>>>,
+) -> Result<JoinHandle<()>, String> {
+    let mut client = MqttClient::new(pub_path, "calypso-encoder");
+    client.connect().expect("Could not connect to Siren!");
+    let rx = client.start_consumer().expect("Could not begin consuming");
+    client
+        .subscribe("Calypso/Bidir/Command/#")
+        .expect("Could not subscribe!");
+
+    // do the default initialization for all, do outside of the thread to save time negotiating when send_can comes up
+    let mut writable_send_map = send_map.write().expect("Could not modify send messages!");
+    for key in ENCODABLE_KEY_LIST {
+        let packet = EncodableMessage::new(String::clone(&key.to_owned()), Vec::new());
+        let ret = packet.encode();
+        writable_send_map.insert(ret.id, ret);
     }
+    drop(writable_send_map);
+
+    let join_handle = thread::spawn(move || {
+        for msg in rx.iter() {
+            if let Some(msg) = msg {
+                let buf = match command_data::CommandData::parse_from_bytes(msg.payload()) {
+                    Ok(buf) => buf,
+                    Err(err) => {
+                        println!("Could not decode command: {}", err);
+                        continue;
+                    }
+                };
+                let key = match msg.topic().split('/').last() {
```

Comment:
we need to be mindful to keep the last topic name unique then

Reply:
Yup since the generated code matches the key to a function to encode, it will not compile if the key is specified twice in the YAML.